src/Symfony/Component/Utf8/Utf8Trait.php

```diff
@@ -0,0 +1,354 @@
+     */
+    public function reverse()
+    {
+        $result = clone $this;
+        $result->string = implode('', array_reverse(iterator_to_array($this)));
+
+        return $result;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function replace($from, $to, int &$count = null)
+    {
+        if (is_string($from)) {
+            $from = array($from);
+        } elseif (!is_array($from)) {
+            throw new InvalidArgumentException('Search pattern must be a valid string or array of strings.');
+        }
+
+        if (is_string($to)) {
+            if (!preg_match('//u', $to)) {
+                throw new InvalidArgumentException(sprintf('Given pattern replacement "%s" is not a valid UTF-8 encoded string.', $to));
+            }
+        } elseif (is_array($to)) {
+            foreach ($to as $replacement) {
+                if (!preg_match('//u', $replacement)) {
+                    throw new InvalidArgumentException(sprintf('Given pattern replacement "%s" is not a valid UTF-8 encoded string.', $replacement));
+                }
+            }
+        } else {
+            throw new InvalidArgumentException('Pattern replacement must be a valid string or array of strings.');
+        }
+
+        foreach ($from as $i => $regex) {
+            if (!preg_match('//u', $regex)) {
+                throw new InvalidArgumentException(sprintf('Given pattern "%s" is not a valid UTF-8 encoded string.', $regex));
+            }
+        }
+
+        $result = clone $this;
+        $result->string = str_replace($from, $to, $this->string, $count);
+
+        return $result;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function replaceIgnoreCase($from, $to, int &$count = null)
+    {
+        if (is_string($from)) {
+            $from = array($from);
+        } elseif (!is_array($from)) {
+            throw new InvalidArgumentException('Search pattern must be a valid string or array of strings.');
+        }
+
+        if (is_string($to)) {
+            if (!preg_match('//u', $to)) {
+                throw new InvalidArgumentException(sprintf('Given pattern replacement "%s" is not a valid UTF-8 encoded string.', $to));
+            }
+        } elseif (is_array($to)) {
+            foreach ($to as $replacement) {
+                if (!preg_match('//u', $replacement)) {
+                    throw new InvalidArgumentException(sprintf('Given pattern replacement "%s" is not a valid UTF-8 encoded string.', $replacement));
+                }
+            }
+        } else {
+            throw new InvalidArgumentException('Pattern replacement must be a valid string or array of strings.');
+        }
+
+        foreach ($from as $i => $regex) {
+            if (!preg_match('//u', $regex)) {
+                throw new InvalidArgumentException(sprintf('Given pattern "%s" is not a valid UTF-8 encoded string.', $regex));
+            }
+
+            $from[$i] = sprintf('/%s/ui', '' === $regex ? '$^' : preg_quote($regex, '/'));
```

Comment:
`$^` never matches anything, or am I wrong? In this case an `unset` might be better, especially since the array might end up being empty at the end and an exception would be nice to indicate incorrect usage.